File: 53-maximum-subarray/maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(numbers) {
    let max = numbers[0];
    let cur = numbers[0];
    for (let i = 1; i < numbers.length; i++) {
        const value = numbers[i];
        if (cur + value >= value) {
            cur += value;
        } else {
            cur = value;
        }

        max = Math.max(cur, max);
    }

    return max;
};

File: 53-maximum-subarray/maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int current = 0, a = INT_MIN;
        for (int i : nums) {
            current += i;
            a = max(a, current);
            if (current < 0) {
                current = 0;
            }
        }
        return a;
    }
};
